### PR TITLE
Add Stripe Form Field & Temporary Submission Saving

### PR DESCRIPTION
**Pull request by Soare-Robert-Daniel**
Closes #.

### Summary

This combines the PRs https://github.com/Codeinwp/otter-blocks/pull/1778 and https://github.com/Codeinwp/otter-blocks/pull/1769

### Screenshots 

image

image

----

### Test instructions

⚠️ You will need an Otter Pro Key and a Stripe account (test mode) with some products.

The flow of the Stripe Field

1. You a Form with some Stripe fields (you can as many as you want)
2. When you press Submit, you will be redirected to the Stripe Checkout Session page (similar to Stripe Checkout Block)
3. Before completing the Card form, go to the Otter Submission page in another browser tab. You will see the post Marked as a `Draft` -- more about this behavior: https://github.com/Codeinwp/otter-blocks/pull/1769.
4. Go back to the Stripe session and complete the payment with some dummy data.
5. After completing the payment, you will be redirected to the same page, but the URL will have a special param called `stripe_checkout`. That is the ID of the session, and it will be used to determine if a user has paid or not.
6. If you see the Success message and go back to Otter Submission, you will see that the submission is no longer a `Draft`

ℹ️ If you enable webhook and autoresponder, while in the paying stage, those should not trigger. They must trigger *ONLY* if the payment is `paid` (you check this in Stripe Dashboard marked as Complete).

⚠️ To confirm the Submission, you must be on the redirect that Stripe will offer to you. If not, the submission will stay as a Draft.

ℹ️ If the payment is marked as done, but you did not visit the special URL. We have a corn job that runs every hour (this value is still in discussion) and checks if there are Submissions in `Draft` mode but with complete payment. If it detects such cases, it will trigger the confirmation process and the remaining functions (webhooks, autoresponder).

You must test if:

- The Stripe Checkout Page has the same products listed in the Form. If not, this is a sync issue, and you will need to provide the test instance.
- The Webhook, autoresponder, etc trigger before the payment is finished.

⚠️ Have in mind the other Form issues that have been reported: #1802 #1804, which will be solved in different PR

---- 

### Checklist before the final review

- [ ] Included E2E or unit tests for the changes in this PR.
- [x] Visual elements are not affected by independent changes.
- [x] It is at least compatible with the [minimum WordPress version](https://wordpress.org/plugins/otter-blocks/).
- [x] It loads additional script in frontend only if it is required.
- [x] Does not impact the [Core Web Vitals](https://web.dev/vitals/).
- [x] In case of deprecation, old blocks are safely migrated.
- [x] It is usable in Widgets and FSE.
- [x] Copy/Paste is working if the attributes are modified.
- [x] PR is following [the best practices]()

### CODE REVIEW

**Comment by Soare-Robert-Daniel** [comment]
@Codeinwp/design-team, here is a short preview of the workflow

https://github.com/Codeinwp/otter-blocks/assets/17597852/bd264f6e-74d0-48ac-9c8d-f52c6165800b

ℹ️  We decided to use the Stripe Checkout session.

When you press submit for the first name, the data from the field are saved into the Submissions table as `Draft`. After the payment has been made, Stripe will redirect you to the same page and Submission will change its status from 'Draft' to `Unread`.

**Comment by mghenciu** [comment]
Thank you for the update Robert.

To summarise, the idea behind this feature is to **redirect the users to a Stripe Checkout form - but also add a new submission in the Otter Submissions list.**

___
From the previous question: 
> @design-team, we need your help with how the Stripe field for Form will look.

are you referring to how should this look in the backend or frontend?


Also regarding the error handling and different states that at Stripe Checkout could end with, how are we managing those. For example let's say the user closes the Stripe Checkout tab, basically abandoning the flow. What would happen to the Otter form?


Thank you!

**Comment by Soare-Robert-Daniel** [comment]
> are you referring to how should this look in the backend or frontend?

Yes,

> Also regarding the error handling and different states that at Stripe Checkout could end with, how are we managing those. For example let's say the user closes the Stripe Checkout tab, basically abandoning the flow. What would happen to the Otter form?

Since we use the Stripe Checkout Session, do not need to handle errors from Stripe. Stripes give us two options: success and cancel. If it is a success, we display the success message of the form. If cancel, we display the error with unpaid purchase. 

Now, if you are in the stripe checkout session, close the tab. Nothing will happen to the user regarding the Form -- the same behavior can be seen in Stripe Checkout Block. But since we save the submission, the submission will remain in the database as `Draft`.  As discussed with @HardeepAsrani, if you see multiple submissions in the `Draft` state, you will have a clue that something is wrong: webhook does not work properly, no autoresponder, or people, can not complete the payment in Stripe.

The Form will _react_ only when it is open in a URL that has the param `stripe_checkout` attached to it -- same behavior as in 
Stripe Checkout Block. And this URL is generated by Stripe at the end of the checkout session on their page (you can see the redirect in the video).

File: inc/plugins/class-options-settings.php [hunk 1]
```diff
@@ -216,6 +216,18 @@ public function register_settings() {
 			)
 		);
 
+		register_setting(
```
Comment (HardeepAsrani) [inline inc/plugins/class-options-settings.php:219]: Just the reminder, we can remove the stuff that we didn't implement like this.

File: plugins/otter-pro/inc/plugins/class-form-pro-features.php [hunk 2]
```diff
@@ -512,6 +513,158 @@ public function replace_magic_tags( $content, $form_inputs ) {
...
+			array(
+				'stripe_checkout' => '{CHECKOUT_SESSION_ID}', // Testing mode.
+			),
+			$form_data->get_data_from_payload( 'postUrl' )
+		);
+
+		$payload['success_url'] = $permalink;
+		$payload['cancel_url']  = $permalink;
+
+		// For @HardeepAsrani: by pre-filling the customer email, it will appear in the Stripe Checkout page without the possibility to change it. You can try it first then we can remove it if you want.
```
Comment (HardeepAsrani) [inline plugins/otter-pro/inc/plugins/class-form-pro-features.php:618]: We can remove pre-filled email as you explained it won't let users change email on checkout page.

File: src/dashboard/components/pages/Integrations.js
Comment (HardeepAsrani) [inline src/dashboard/components/pages/Integrations.js:1]: We can revert these changes.

File: inc/server/class-form-server.php [hunk 3]
```diff
@@ -176,6 +180,36 @@ public function register_routes() {
...
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'confirm_form_submission' ),
+					'permission_callback' => function ( $request ) {
+						$session = $request->get_param( 'stripe_checkout' );
+
+						if ( apply_filters( 'otter_form_session_confirmation', $session ) ) {
+							return __return_true();
+						}
+						return __return_false();
```
Comment (HardeepAsrani) [inline inc/server/class-form-server.php:196]: NIT: you can just return a false. 😄

File: plugins/otter-pro/inc/plugins/class-form-emails-storing.php [hunk 4]
```diff
@@ -67,6 +71,13 @@ public function init() {
 		add_action( 'add_meta_boxes', array( $this, 'add_form_record_meta_box' ) );
 		add_action( 'admin_menu', array( $this, 'handle_admin_menu' ) );
 		add_action( 'save_post', array( $this, 'form_record_save_meta_box' ), 10, 2 );
+
+		add_filter( 'otter_form_record_confirm', array( $this, 'confirm_submission' ), 10, 2 );
+
+		add_action( 'transition_post_status', array( $this, 'apply_hooks_on_draft_transition' ), 10, 3 );
```
Comment (HardeepAsrani) [inline plugins/otter-pro/inc/plugins/class-form-emails-storing.php:77]: Not needed but you can also use it like this to target specific status:  `add_action( "draft_to_publish", 'do_stuff` );

File: plugins/otter-pro/inc/plugins/class-form-emails-storing.php [hunk 5]
```diff
@@ -207,19 +222,23 @@ public function store_form_record( $form_data ) {
 		$meta = array(
 			'form'     => array(
 				'label' => 'Form',
```
Comment (HardeepAsrani) [inline plugins/otter-pro/inc/plugins/class-form-emails-storing.php:224]: All the labels here, we need to wrap in i18n funcs

**Review by HardeepAsrani — APPROVED** [review]
I've left few nits here and there but overall looks good. We need to do a through QA here to confirm the new and the existing features work as before with and without the Pro being installed.

**Comment by irinelenache** [comment]
@Soare-Robert-Daniel After adding 2 Stripe fields on an [existing form](https://vertis.d.pr/i/jw3ylP) (from Web agency starter site), i got a fatal error when i opened the page in frontend 
```
Warning:  require(/s3-mole-fixes/wordpress/wp-content/plugins/otter-pro/inc/css/blocks/class-form-stripe-field-css.php): failed to open stream: No such file or directory in /s3-mole-fixes/wordpress/wp-content/plugins/otter-pro/autoloader.php on line 148
Fatal error:  require(): Failed opening required '/s3-mole-fixes/wordpress/wp-content/plugins/otter-pro/inc/css/blocks/class-form-stripe-field-css.php' (include_path='.:/usr/share/php') in /s3-mole-fixes/wordpress/wp-content/plugins/otter-pro/autoloader.php on line 148
```

**Comment by Soare-Robert-Daniel** [comment]
@irinelenache, I updated your instance.

⚠️ It seams that your stripe account does not allow the creation of checkout session. I replaced it with mine.

**Comment by irinelenache** [comment]
@Soare-Robert-Daniel 

- [ ] The **redirect on submit** feature doesn't work after finishing the Stripe payment https://vertis.d.pr/v/dGPZiL
- [ ] Is it expected to have this many cron events tied to the form? https://vertis.d.pr/v/SW66ov I tried to delete them all and got about 6 new ones in 2 minutes
- [ ] I tried to manually run all cron events related to the form and the submission was still in draft, also the post submission events didn't happen

You can check here:
```
 Admin area URL: https://mole-fixes.s3-tastewp.com/wp-admin 
 Username: irinel 
 Password: jButlSpxvLQ 
```

**Comment by irinelenache** [comment]
@Soare-Robert-Daniel Tested the latest changes and everything is fine now 👍

**Comment by Soare-Robert-Daniel** [comment]
ℹ️ Now, redirect behavior will be switching the link in the same tab rather than opening a new one.

**Comment by irinelenache** [comment]
@Soare-Robert-Daniel Had another look regarding the issue discussed on Slack and it's fixed now, i'll move this to **Ready to merge** ✅

File: .fleet/settings.json [hunk 6]
```diff
@@ -0,0 +1,3 @@
+{
+    "toolchains": []
```
Comment (HardeepAsrani) [inline .fleet/settings.json:2]: Can we remove this?

Comment (HardeepAsrani) [reply .fleet/settings.json:2]: @Soare-Robert-Daniel

Comment (Soare-Robert-Daniel) [reply .fleet/settings.json:2]: Yes. And better put it to `.gitignore` since it is IDE-related.

Comment (HardeepAsrani) [reply .fleet/settings.json:2]: Can you make the changes? Delete it and put it in gitignore?
